lib/pengine/pe_output.c

```diff
@@ -0,0 +1,283 @@
+    bool multi_state = va_arg(args, int);
+    bool unique = va_arg(args, int);
+    bool managed = va_arg(args, int);
+    bool failed = va_arg(args, int);
+    bool failure_ignored = va_arg(args, int);
+    const char *target_role = va_arg(args, const char *);
+    long options = va_arg(args, long);
+    resource_t *rsc = va_arg(args, resource_t *);
+
+    xml_private_t *priv = out->priv;
+    GListPtr gIter = rsc->children;
+
+    int rc = pe__call_message(out, true, fn, "clone", 7
+                            , "id", id
+                            , "multi_state", BOOL2STR(multi_state)
+                            , "unique", BOOL2STR(unique)
+                            , "managed", BOOL2STR(managed)
+                            , "failed", BOOL2STR(failed)
+                            , "failure_ignored", BOOL2STR(failure_ignored)
+                            , "target_role", target_role);
+    CRM_ASSERT(rc == 0);
+
+    for (; gIter != NULL; gIter = gIter->next) {
+        resource_t *child_rsc = (resource_t *) gIter->data;
+
+        child_rsc->fns->output(child_rsc, options, out);
+    }
+
+    g_queue_pop_tail(priv->parent_q); // clone
+    return rc;
+}
+
+static int
+pe__group(pcmk__output_t *out, pe__name_and_nvpair_message_fn_t fn, va_list args)
+{
+    long options = va_arg(args, long);
+    resource_t *rsc = va_arg(args, resource_t *);
+
+    GListPtr gIter = rsc->children;
+    char *count = crm_itoa(g_list_length(gIter));
+    xml_private_t *priv = out->priv;
+
+    int rc = pe__call_message(out, true, fn, "group", 2
+                              , "id", rsc->id
+                              , "number_resources", count);
+    free(count);
+    CRM_ASSERT(rc == 0);
+    CRM_ASSERT(priv != NULL);
+
+    for (; gIter != NULL; gIter = gIter->next) {
+        resource_t *child_rsc = (resource_t *) gIter->data;
+
+        child_rsc->fns->output(child_rsc, options, out);
+    }
+
+    g_queue_pop_tail(priv->parent_q); // group
+    return rc;
+}
+
+static int
+pe__resource(pcmk__output_t *out, pe__name_and_nvpair_message_fn_t fn, va_list args)
+{
+    long options = va_arg(args, int);
+    long is_print_pending = options & pe_print_pending;
+    long is_print_dev = options & pe_print_dev;
+
+    const char *id = va_arg(args, const char *);
+    const char *ra_name = va_arg(args, const char *);
+    const char *rsc_state = va_arg(args, const char *);
+    const char *target_role = va_arg(args, const char *);
+    bool active = va_arg(args, int);
+    bool orphaned = va_arg(args, int);
+    bool blocked = va_arg(args, int);
+    bool managed = va_arg(args, int);
+    bool failed = va_arg(args, int);
+    bool failure_ignored = va_arg(args, int);
+    char *nodes_running_on = crm_itoa(va_arg(args, int));
+    const char *pending = va_arg(args, const char*);
+    bool provisional  = va_arg(args, int);
+    bool runnable = va_arg(args, int);
+    char *priority = crm_ftoa(va_arg(args, double));
```

Comment:
I'm pretty sure rsc->priority is int, where is the double coming from?

Reply:
There is an equivalent line `status_print("priority=\"%f\" ", (double)rsc->priority);` in the `native_print_xml` function (lib/pengine/native:467 in the master branch), so I thought, that I should also use the float type.

Reply:
Huh, that's interesting. There's no indication in git history why that was chosen. It is a good idea to keep backward compatibility for this PR. I'll make a note to look into it later.